[Messenger] Multiples buses, scoping handlers per bus, PSR-4 discovery & debug (#10689)
symfony/symfony-docs: messenger/multiple_buses.rst

.. index::
    single: Messenger; Multiple buses

Multiple Buses
==============

> A common architecture when building application is to separate commands from

Reviewer: ```suggestion
A common architecture when building applications is to separate commands from
```

queries. Commands are actions that do something and queries fetch data. This
is called CQRS (Command Query Responsibility Segregation). See Martin Fowler's
`article about CQRS`_ to learn more. This architecture could be used together
with the Messenger component by defining multiple buses.

> A **command bus** is a little different from a **query bus**. As example, command

Reviewer: ```suggestion
A **command bus** is a little different from a **query bus**. For example, command
```
or "As an example"

buses usually don't provide any results and query buses are rarely asynchronous.

> You can configure these buses and their rules by using middleware.

Reviewer: ```suggestion
You can configure these buses and their rules by using middlewares.
```

Reviewer: Just like software and hardware, makes sense 👍

                'buses' => array(
                    'messenger.bus.commands' => array(
                        'middleware' => array(
                            'validation',
                            'doctrine_transaction',
                        ),
                    ),
                    'messenger.bus.queries' => array(
                        'middleware' => array(
                            'validation',
                        ),
                    ),
                    'messenger.bus.events' => array(
                        'middleware' => array(
                            'validation',
                            'allow_no_handler',
                        ),
                    ),
                ),
            ),
        ));

This will generate the ``messenger.bus.commands``, ``messenger.bus.queries``
and ``messenger.bus.events`` services that you can inject in your services.

Type-hints and Auto-wiring
--------------------------

Auto-wiring is a great feature that allows you to reduce the amount of configuration
required for your service container to be created. When using multiple buses, by default,

> the auto-wiring will not work as it won't know which bus to inject in your own services.

Reviewer 2: That's not totally true, right? The `messenger.bus.commands` service will be sent to you if you use the type-hint, by default. We should probably mention this.

> In order to clarify this, you can use the DependencyInjection's binding capabilities

Reviewer: ```suggestion
In order to clarify this, you can use the ``DependencyInjection`` component's binding capabilities
```

or the "dependency injection component's…"

to clarify which bus will be injected based on the argument's name:

.. configuration-block::

    .. code-block:: yaml

        # config/services.yaml
        services:
            _defaults:
                # ...

                bind:
                    $commandBus: '@messenger.bus.commands'
                    $queryBus: '@messenger.bus.queries'

>                     $eventBus: '@messenger.bus.events'

Reviewer 2: Just a note - we'll need another PR in Symfony 4.2, because (I believe) these bindings are done automatically.

    .. code-block:: xml

        <!-- config/services.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                http://symfony.com/schema/dic/services/services-1.0.xsd">

            <services>
                <defaults>
                   <bind key="$commandBus" type="service" id="messenger.bus.commands" />
                   <bind key="$queryBus" type="service" id="messenger.bus.queries" />
                   <bind key="$eventBus" type="service" id="messenger.bus.events" />
                </defaults>
            </services>
        </container>

    .. code-block:: php

        // config/services.php

        $container->bind('$commandBus', 'messenger.bus.commands');
        $container->bind('$queryBus', 'messenger.bus.queries');
        $container->bind('$eventBus', 'messenger.bus.events');

Restrict handlers per bus & PSR-4 discovery
-------------------------------------------

By using the ``messenger.message_handler`` tag on a service, you're declaring

> it as a handler. But without more attributes, the it would actually be available

Reviewer 3: "the it would" => "it would"

any bus. As you don't want to dispatch a message to the wrong bus without notice,

> you can use the ``bus``  attribute with the expected bus name as value:

Reviewer 2: It's a bit confusing because most users won't even be aware of this tag - it'll just be autoconfigured for them. So maybe something like:

```
By default, each handler class will be available to handle messages on *all* of your busses. To prevent
dispatching a message to the wrong bus without an error, you can restrict each handler to a specific
bus using the `messenger.message_handler` tag:
```

        # config/services.yaml
        services:
            App\MessageHandler\SomeCommandHandler:
                tags: [{ name: messenger.message_handler, bus: messenger.bus.commands }]

    .. code-block:: xml

        <!-- config/services.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                http://symfony.com/schema/dic/services/services-1.0.xsd">

            <services>
                <service id="App\MessageHandler\SomeCommandHandler">
                    <tag name="messenger.message_handler" bus="messenger.bus.commands" />
                </service>
            </services>
        </container>

    .. code-block:: php

        // config/services.php

        $container->services()
            ->set(App\MessageHandler\SomeCommandHandler::class)
            ->tag('messenger.message_handler', ['bus' => 'messenger.bus.commands']);

This way, the ``App\MessageHandler\SomeCommandHandler`` handler will only be
known by the ``messenger.bus.commands`` bus.

> You can also leverage PSR-4 discovery to configure your handlers automatically:

Reviewer 2: I don't think people really know this is called PSR-4 discovery. How about:

> You can also automatically add this tag to a number of classes by following a naming convention
> and registering all of the handler services by name with the correct tag:

.. configuration-block::

    .. code-block:: yaml

>         command_handlers:

Reviewer 2: ```
# config/services.yaml

    # put this after the `App\` line that registers all your services
```

>             namespace: App\MessageHandler\

Reviewer 2: To follow how it's normally done, should we just use this as the key above? And then not have this `namespace` config key?

            tags:
                - { name: messenger.message_handler, bus: messenger.bus.queries }

    .. code-block:: xml

        <!-- config/services.xml -->
        <?xml version="1.0" encoding="UTF-8" ?>
        <container xmlns="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                http://symfony.com/schema/dic/services/services-1.0.xsd">

            <services>
                <!-- command handlers -->
                <prototype namespace="App\MessageHandler\" resource="%kernel.project_dir%/src/MessageHandler/*CommandHandler.php">
                    <tag name="messenger.message_handler" bus="messenger.bus.commands" />
                </service>
                <!-- query handlers -->
                <prototype namespace="App\MessageHandler\" resource="%kernel.project_dir%/src/MessageHandler/*QueryHandler.php">
                    <tag name="messenger.message_handler" bus="messenger.bus.queries" />
                </service>
            </services>
        </container>

    .. code-block:: php

>         // config/services.php

Reviewer 2: Extra line break

        // Command handlers
        $container->services()
            ->load('App\MessageHandler\\', '%kernel.project_dir%/src/MessageHandler/*CommandHandler.php')
            ->tag('messenger.message_handler', ['bus' => 'messenger.bus.commands']);

        // Query handlers
        $container->services()
            ->load('App\MessageHandler\\', '%kernel.project_dir%/src/MessageHandler/*QueryHandler.php')
            ->tag('messenger.message_handler', ['bus' => 'messenger.bus.queries']);

Debug buses

> -----------

Reviewer 2: Debugging the Buses

> A ``debug:messenger`` command will list available messages & handlers per bus.

Reviewer 3: Use the present tense, i.e. "The debug:messanger command lists ..."
